packages/loopring_v3/contracts/amm/libamm/AmmBlockReceiver.sol

```diff
@@ -4,124 +4,135 @@ pragma solidity ^0.7.0;
     using AmmJoinProcess    for AmmData.State;
+    using AmmPoolToken      for AmmData.State;
     using AmmUpdateProcess  for AmmData.State;
     using BlockReader       for ExchangeData.Block;
+    using MathUint          for uint;
 
-    function beforeBlockSubmitted(
+    function beforeBlockSubmission(
         AmmData.State      storage S,
         ExchangeData.Block memory  _block,
-        uint                       txIdx,
-        bytes              memory  auxiliaryData
+        bytes              memory  poolTxData,
+        uint                       txIdx
         )
         public
         returns (uint)
     {
-        AmmData.PoolTransaction[] memory poolTransactions = abi.decode(
-            auxiliaryData,
-            (AmmData.PoolTransaction[])
-        );
-
-        // Cache the domain seperator to save on SLOADs each time it is accessed.
-        uint size = S.tokens.length;
-        AmmData.Context memory ctx = AmmData.Context({
-            _block: _block,
-            exchange: S.exchange,
-            exchangeDepositContract: address(S.exchange.getDepositContract()),
-            txIdx: txIdx,
-            domainSeparator: S.domainSeparator,
-            exchangeDomainSeparator: S.exchange.getDomainSeparator(),
-            ammActualL2Balances: new uint96[](size),
-            ammExpectedL2Balances: new uint96[](size),
-            numTransactionsConsumed: 0,
-            tokens: S.tokens,
-            size: size,
-            poolTokenBase: AmmData.LP_TOKEN_BASE(),
-            poolTokenInitialSupply: AmmData.LP_TOKEN_INITIAL_SUPPLY()
-        });
+        AmmData.Context memory ctx = getContext(S, _block, txIdx);
+
+        // Question(brecht): is it OK not to check the following for every poolTx?
 
         BlockReader.BlockHeader memory header = _block.readHeader();
         require(header.exchange == address(ctx.exchange), "INVALID_EXCHANGE");
 
-        // The openning AMM updates
-        // This also pulls the AMM balances onchain.
-        S.processAmmUpdates(ctx, true);
+        // The openning AMM updates will pull the latest AMM layer2 balances onchain.
+        S.approveAmmUpdates(ctx, true);
 
-        // Process all pool transactions
-        for (uint i = 0; i < poolTransactions.length; i++) {
-            _processPoolTransaction(S, ctx, poolTransactions[i]);
-        }
+        processPoolTx(S, ctx, poolTxData);
 
-        // Deposit/Withdraw to/from the AMM account when necessary
-        for (uint i = 0; i < size; i++) {
-            _processPoolBalance(
-                S,
-                ctx,
-                ctx.tokens[i],
-                ctx.ammExpectedL2Balances[i],
-                ctx.ammActualL2Balances[i]
-            );
-        }
+        appproveDepositOrWithdrawal(ctx);
 
-        // The closing AMM updates
-        S.processAmmUpdates(ctx, false);
+        S.approveAmmUpdates(ctx, false);
 
         return ctx.numTransactionsConsumed;
     }
```

Comment:
Small improvement/simplication I didn't think of before: We don't have to track `numTransactionsConsumed`, we can just calculate it here as `ctx.txIdx - txIdx`.